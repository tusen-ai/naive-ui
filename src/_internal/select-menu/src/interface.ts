import type { VNodeChild, Ref, UnwrapRef, VNode, HTMLAttributes } from 'vue'
import type { TreeNode } from 'treemate'
import type {
  SelectBaseOption,
  SelectGroupOption,
  SelectIgnoredOption
} from '../../../select/src/interface'
import { createInjectionKey } from '../../../_utils/vue/create-injection-key'

export type Size = 'small' | 'medium' | 'large' | 'huge'

export type RenderLabel = (
  option: SelectBaseOption & SelectGroupOption & SelectIgnoredOption,
  selected: boolean
) => VNodeChild

export type RenderLabelImpl = (
  option: SelectBaseOption | SelectGroupOption | SelectIgnoredOption,
  selected: boolean
) => VNodeChild

export type RenderOption = (info: {
  node: VNode
  option: SelectBaseOption & SelectGroupOption & SelectIgnoredOption
  selected: boolean
}) => VNodeChild

export type RenderOptionImpl = (info: {
  node: VNode
  option: SelectBaseOption | SelectGroupOption | SelectIgnoredOption
  selected: boolean
}) => VNodeChild

export type NodeProps = (
  option: SelectBaseOption | SelectGroupOption
) => HTMLAttributes & Record<string, unknown>

export interface InternalSelectMenuInjection {
  handleOptionMouseEnter: (
    e: MouseEvent,
    tmNode: TreeNode<SelectBaseOption>
  ) => void
  handleOptionClick: (e: MouseEvent, tmNode: TreeNode<SelectBaseOption>) => void
  showCheckmarkRef: Ref<boolean>
  valueSetRef: Ref<Set<number | string>>
  pendingTmNodeRef: Ref<TreeNode<SelectBaseOption> | null>
  multipleRef: Ref<boolean>
  valueRef: Ref<string | number | Array<string | number> | null>
  renderLabelRef: Ref<RenderLabel | undefined>
  renderOptionRef: Ref<RenderOption | undefined>
  labelFieldRef: Ref<string>
  valueFieldRef: Ref<string>
  nodePropsRef: Ref<NodeProps | undefined>
}

export interface InternalExposedProps {
  selfRef: Ref<HTMLElement | null>
  getPendingTmNode: () => TreeNode<SelectBaseOption> | null
  prev: () => void
  next: () => void
}

export const internalSelectionMenuInjectionKey =
  createInjectionKey<InternalSelectMenuInjection>('n-internal-select-menu')

export const internalSelectionMenuBodyInjectionKey = createInjectionKey<
Ref<HTMLElement | null>
>('n-internal-select-menu-body')

export type InternalSelectMenuRef = UnwrapRef<InternalExposedProps>
