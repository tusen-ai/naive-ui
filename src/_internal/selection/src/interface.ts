import { type VNodeChild } from 'vue'
import type { SelectOption } from '../../../select/src/interface'

export type RenderTag = (props: {
  option: SelectOption
  handleClose: () => void
}) => VNodeChild
