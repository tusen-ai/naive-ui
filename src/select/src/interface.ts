import { type TreeMate } from 'treemate'
import { type CSSProperties, type VNodeChild, type VNode } from 'vue'

export type SelectMixedOption =
  | SelectBaseOption
  | SelectGroupOption
  | SelectIgnoredOption

export interface SelectBaseOption<
  V = string | number,
  L = string | ((option: SelectBaseOption<V>, selected: boolean) => VNodeChild)
> {
  value?: V
  label?: L
  class?: string
  style?: string | CSSProperties
  disabled?: boolean
  render?: (info: {
    node: VNode
    option: SelectBaseOption<V>
    selected: boolean
  }) => VNodeChild
  [k: string]: unknown
}

export interface SelectGroupOptionBase {
  label?: string | ((option: SelectGroupOption) => VNodeChild)
  type: 'group'
  children?: SelectBaseOption[]
  render?: (info: { node: VNode, option: SelectGroupOption }) => VNodeChild
  [k: string]: unknown
}

export interface SelectIgnoredOption {
  type: 'ignored'
  value: string | number
  [k: string]: unknown
}

export type ValueAtom = string | number
export type Value = ValueAtom | string[] | number[] | ValueAtom[]
export type OnUpdateValue = (
  value: string &
  number &
  ValueAtom &
  string[] &
  number[] &
  ValueAtom[] &
  (ValueAtom | null) &
  (string[] | null) &
  (number[] | null) &
  (ValueAtom[] | null),
  option: SelectBaseOption & null & SelectBaseOption[]
) => void
export type OnUpdateValueImpl = (
  value:
  | ValueAtom
  | string[]
  | number[]
  | ValueAtom[]
  | (ValueAtom | null)
  | (string[] | null)
  | (number[] | null)
  | (ValueAtom[] | null),
  option: SelectBaseOption | null | SelectBaseOption[]
) => void
export type SelectTreeMate = TreeMate<
SelectBaseOption,
SelectGroupOption,
SelectIgnoredOption
>

export type Size = 'tiny' | 'small' | 'medium' | 'large'

// Public interfaces
export type SelectOption = SelectBaseOption<string | number>
export type SelectGroupOption =
  | (SelectGroupOptionBase & {
    /** @deprecated should use key and label instead */
    name?: string
  })
  | (SelectGroupOptionBase & {
    key: string | number
  })

export interface SelectInst {
  focus: () => void
  blur: () => void
}

export type SelectFallbackOption = (value: string & number) => SelectOption
export type SelectFallbackOptionImpl = (value: string | number) => SelectOption
export type SelectFilter = (pattern: string, option: SelectOption) => boolean
