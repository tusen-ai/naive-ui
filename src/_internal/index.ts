export { default as NIconSwitchTransition } from './icon-switch-transition'
export { default as NFadeInExpandTransition } from './fade-in-expand-transition'
export { NBaseClose } from './close'
export { NBaseIcon } from './icon'
export { default as NBaseFocusDetector } from './focus-detector'
export { default as NBaseLoading } from './loading'
export type { BaseLoadingExposedProps } from './loading'
export { default as NInternalSelectMenu } from './select-menu'
export type { InternalSelectMenuRef } from './select-menu'
export { default as NBaseWave } from './wave'
export type { BaseWaveRef } from './wave'
export { default as NBaseMenuMask } from './menu-mask'
export { default as NInternalSelection } from './selection'
export type { InternalSelectionInst } from './selection'
export { default as NBaseSlotMachine } from './slot-machine'
export { default as NBaseClear } from './clear'
export { default as NBaseSuffix } from './suffix'
export { NScrollbar, NxScrollbar } from './scrollbar'
export type { ScrollbarInst, ScrollbarProps } from './scrollbar'
