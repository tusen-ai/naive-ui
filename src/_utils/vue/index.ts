export { getSlot } from './get-slot'
export { getVNodeChildren } from './get-v-node-children'
export { keep } from './keep'
export { omit } from './omit'
export { flatten } from './flatten'
export { call } from './call'
export { keysOf } from './keysOf'
export { render } from './render'
export { getFirstSlotVNode } from './get-first-slot-vnode'
export { createDataKey } from './create-data-key'
export { createRefSetter } from './create-ref-setter'
export { createInjectionKey } from './create-injection-key'
export {
  resolveSlot,
  resolveWrappedSlot,
  resolveSlotWithProps,
  resolveWrappedSlotWithProps,
  isSlotEmpty
} from './resolve-slot'
export type { MaybeArray } from './call'
export { mergeEventHandlers } from './merge-handlers'
export { isNodeVShowFalse } from './is-node-v-show-false'
export { Wrapper } from './wrapper'
