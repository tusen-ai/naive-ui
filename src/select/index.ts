export type {
  NodeProps as SelectNodeProps,
  RenderLabelImpl as SelectRenderLabel,
  RenderOptionImpl as SelectRenderOption
} from '../_internal/select-menu/src/interface'
export type { RenderTag as SelectRenderTag } from '../_internal/selection/src/interface'
export type {
  SelectFilter,
  SelectGroupOption,
  SelectInst,
  SelectOption
} from './src/interface'
export { default as NSelect, selectProps } from './src/Select'
export type { SelectProps, SelectSlots } from './src/Select'
