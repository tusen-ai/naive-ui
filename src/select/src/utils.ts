import { type TreeMateOptions } from 'treemate'
import type {
  SelectBaseOption,
  SelectGroupOption,
  SelectIgnoredOption,
  SelectMixedOption,
  SelectOption
} from './interface'

export function getIsGroup (option: SelectMixedOption): boolean {
  return option.type === 'group'
}

export function getIgnored (option: SelectMixedOption): boolean {
  return option.type === 'ignored'
}

export function patternMatched (pattern: string, value: string): boolean {
  try {
    return !!(
      1 + value.toString().toLowerCase().indexOf(pattern.trim().toLowerCase())
    )
  } catch (err) {
    return false
  }
}

export function createTmOptions (
  valueField: string,
  childrenField: string
): TreeMateOptions<SelectBaseOption, SelectGroupOption, SelectIgnoredOption> {
  const options: TreeMateOptions<
  SelectBaseOption,
  SelectGroupOption,
  SelectIgnoredOption
  > = {
    getIsGroup,
    getIgnored,
    getKey (option: SelectMixedOption): string | number {
      if (getIsGroup(option)) {
        return (
          ((option as SelectGroupOption).name as any) ||
          (option as SelectGroupOption).key ||
          'key-required'
        )
      }
      // Required for non-custom label & value field
      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
      return (option as SelectBaseOption | SelectIgnoredOption)[
        valueField
      ] as NonNullable<SelectOption['value']>
    },
    getChildren (option) {
      return option[childrenField] as SelectBaseOption[]
    }
  }
  return options
}

export function filterOptions (
  originalOpts: SelectMixedOption[],
  filter: (pattern: string, option: SelectBaseOption) => boolean,
  pattern: string,
  childrenField: string
): SelectMixedOption[] {
  if (!filter) return originalOpts
  function traverse (options: SelectMixedOption[]): SelectMixedOption[] {
    if (!Array.isArray(options)) return []
    const filteredOptions: SelectMixedOption[] = []
    for (const option of options) {
      if (getIsGroup(option)) {
        const children = traverse(option[childrenField] as SelectBaseOption[])
        if (children.length) {
          filteredOptions.push(
            Object.assign({}, option, {
              [childrenField]: children
            })
          )
        }
      } else if (getIgnored(option)) {
        continue
      } else if (filter(pattern, option as SelectBaseOption)) {
        filteredOptions.push(option)
      }
    }
    return filteredOptions
  }
  return traverse(originalOpts)
}

export function createValOptMap (
  options: SelectMixedOption[],
  valueField: string,
  childrenField: string
): Map<string | number, SelectBaseOption> {
  const valOptMap = new Map<string | number, SelectBaseOption>()
  options.forEach((option) => {
    if (getIsGroup(option)) {
      ;(option[childrenField] as SelectBaseOption[]).forEach(
        (selectGroupOption: SelectBaseOption) => {
          valOptMap.set(
            selectGroupOption[valueField] as NonNullable<
            SelectBaseOption['value']
            >,
            selectGroupOption
          )
        }
      )
    } else {
      valOptMap.set(
        option[valueField] as NonNullable<SelectBaseOption['value']>,
        option as SelectBaseOption
      )
    }
  })
  return valOptMap
}
