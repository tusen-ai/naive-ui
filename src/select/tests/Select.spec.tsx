import { h, nextTick, type VNode } from 'vue'
import { mount } from '@vue/test-utils'
import {
  NSelect,
  type SelectProps,
  type SelectOption,
  type SelectGroupOption
} from '../index'
import { NInternalSelection, NInternalSelectMenu } from '../../_internal'
import { NTag } from '../../tag'
import { type SelectBaseOption } from '../src/interface'

describe('n-select', () => {
  it('should work with import on demand', () => {
    mount(NSelect)
  })
  it('show menu when trigger clicked', async () => {
    const wrapper = mount(NSelect)
    const inputWrapper = wrapper.findComponent(NInternalSelection)
    expect(wrapper.findComponent(NInternalSelectMenu).exists()).toEqual(false)
    await inputWrapper.trigger('click')
    expect(wrapper.findComponent(NInternalSelectMenu).isVisible()).toEqual(true)
    await inputWrapper.trigger('click')
    expect(wrapper.findComponent(NInternalSelectMenu).isVisible()).toEqual(
      false
    )
    wrapper.unmount()
  })
  it('props.show', () => {
    const wrapper = mount(NSelect, {
      props: {
        show: true
      }
    })
    expect(wrapper.findComponent(NInternalSelectMenu).exists()).toEqual(true)
    wrapper.unmount()
  })
  describe('props.option', () => {
    it('has correct type', () => {
      const options: SelectProps['options'] = [
        {
          label: 'cool1',
          value: 'cool1'
        },
        {
          type: 'group',
          label: 'cool',
          key: 'group cool',
          children: [
            {
              label: 'cool2',
              value: 'cool2'
            }
          ]
        }
      ]
      mount(() => <NSelect options={options} />).unmount()
    })
    it('option.label as render function', () => {
      const options: SelectProps['options'] = [
        {
          label: () => 'cool1+1',
          value: 'cool1'
        },
        {
          type: 'group',
          label: () => 'cool1+2',
          key: 'group cool',
          children: [
            {
              label: () => 'cool1+3',
              value: 'cool2'
            }
          ]
        }
      ]
      const wrapper = mount(NSelect, {
        props: {
          options,
          show: true,
          virtualScroll: false
        }
      })
      const menuWrapper = wrapper.findComponent(NInternalSelectMenu)
      expect(
        ['cool1+1', 'cool1+2', 'cool1+3'].every((label) =>
          menuWrapper.text().includes(label)
        )
      ).toEqual(true)
      wrapper.unmount()
    })
    it('option.render', () => {
      const options: SelectProps['options'] = [
        {
          label: 'cool1',
          value: 'cool1',
          render: ({ node, option }: { node: VNode, option: SelectOption }) => {
            expect(option.label).toEqual('cool1')
            return <div class="cool1">{node}</div>
          }
        },
        {
          type: 'group',
          label: 'cool2',
          key: 'group cool',
          render: ({
            node,
            option
          }: {
            node: VNode
            option: SelectGroupOption
          }) => {
            expect(option.label).toEqual('cool2')
            return <div class="cool2">{node}</div>
          },
          children: [
            {
              label: 'cool3',
              value: 'cool3',
              render: ({ node, option }) => {
                expect(option.label).toEqual('cool3')
                return <div class="cool3">{node}</div>
              }
            }
          ]
        }
      ]
      const wrapper = mount(NSelect, {
        props: {
          options,
          show: true,
          virtualScroll: false
        }
      })
      const menuWrapper = wrapper.findComponent(NInternalSelectMenu)
      expect(menuWrapper.find('.cool1').exists()).toEqual(true)
      expect(menuWrapper.find('.cool2').exists()).toEqual(true)
      expect(menuWrapper.find('.cool3').exists()).toEqual(true)
      wrapper.unmount()
    })
    it('props.renderOption', () => {
      const renderOption: SelectProps['renderOption'] = ({
        node,
        option
      }: {
        node: VNode
        option: SelectOption | SelectGroupOption
      }) => <div class={option.label}>{node}</div>
      const options: SelectProps['options'] = [
        {
          label: 'cool1',
          value: 'cool1'
        },
        {
          type: 'group',
          label: 'cool2',
          key: 'group cool',
          children: [
            {
              label: 'cool3',
              value: 'cool3',
              render: ({ node, option }) => {
                expect(option.label).toEqual('cool3')
                return <div class="cool3">{node}</div>
              }
            }
          ]
        }
      ]
      const wrapper = mount(NSelect, {
        props: {
          options,
          show: true,
          virtualScroll: false,
          renderOption
        }
      })
      const menuWrapper = wrapper.findComponent(NInternalSelectMenu)
      expect(menuWrapper.find('.cool1').exists()).toEqual(true)
      expect(menuWrapper.find('.cool2').exists()).toEqual(true)
      expect(menuWrapper.find('.cool3').exists()).toEqual(true)
      wrapper.unmount()
    })
  })

  it('should work with `render-tag` prop', async () => {
    const options = [
      {
        label: 'test',
        value: 'test',
        type: 'success'
      }
    ]

    const wrapper = mount(NSelect, {
      props: {
        defaultValue: ['test'],
        options,
        multiple: true,
        virtualScroll: false,
        renderTag: ({
          option,
          handleClose
        }: {
          option: SelectBaseOption
          handleClose: () => void
        }) => {
          return h(
            NTag,
            {
              type: option.type as 'success',
              closable: true,
              onClose: handleClose
            },
            { default: () => option.label }
          )
        }
      }
    })

    expect(wrapper.find('.n-base-selection-tag-wrapper').exists()).toBe(true)
    expect(wrapper.findComponent(NTag).exists()).toBe(true)
    expect(wrapper.findComponent(NTag).props('type')).toContain('success')
    await wrapper.find('.n-tag__close').trigger('click')
    expect(wrapper.findComponent(NTag).exists()).toBe(false)
    wrapper.unmount()
  })

  it('should work with `render-label` prop', async () => {
    const options = [
      {
        label: 'test',
        value: 'test'
      }
    ]

    const wrapper = mount(NSelect, {
      attachTo: document.body,
      props: {
        defaultValue: 'test',
        options,
        virtualScroll: false,
        renderLabel: (option: SelectOption) => {
          return 'render-' + String(option.label)
        }
      }
    })

    expect(wrapper.find('.n-base-selection-input').text()).toBe('render-test')
    await wrapper.setProps({ show: true })
    await nextTick()
    expect(
      document.querySelector('.n-base-select-option--selected')?.innerHTML
    ).toContain('render-test')
    wrapper.unmount()
  })

  it('should work with `disabled` prop', async () => {
    const wrapper = mount(NSelect)

    expect(wrapper.find('.n-base-selection').classes()).not.toContain(
      'n-base-selection--disabled'
    )
    await wrapper.setProps({
      disabled: true
    })
    expect(wrapper.find('.n-base-selection').classes()).toContain(
      'n-base-selection--disabled'
    )
    wrapper.unmount()
  })

  it('should work with `filterable` prop', async () => {
    const wrapper = mount(NSelect)

    expect(wrapper.find('input').exists()).not.toBe(true)
    expect(wrapper.find('.n-base-selection-input').exists()).not.toBe(true)
    await wrapper.setProps({
      filterable: true
    })
    expect(wrapper.find('input').exists()).toBe(true)
    expect(wrapper.find('.n-base-selection-input').exists()).toBe(true)
    wrapper.unmount()
  })

  it('should work with `loading` prop', async () => {
    const wrapper = mount(NSelect)

    expect(wrapper.find('.n-base-loading__container').exists()).not.toBe(true)
    await wrapper.setProps({
      loading: true
    })
    expect(wrapper.find('.n-base-loading__container').exists()).toBe(true)
    wrapper.unmount()
  })

  it('should work with `multiple` prop', async () => {
    const wrapper = mount(NSelect)

    expect(wrapper.find('.n-base-selection').classes()).not.toContain(
      'n-base-selection--multiple'
    )
    await wrapper.setProps({
      multiple: true
    })
    expect(wrapper.find('.n-base-selection').classes()).toContain(
      'n-base-selection--multiple'
    )
    wrapper.unmount()
  })

  it('should work with `placeholder` prop', async () => {
    const wrapper = mount(NSelect, {
      props: {
        placeholder: 'test'
      }
    })
    expect(wrapper.find('.n-base-selection-placeholder').text()).toBe('test')
    wrapper.unmount()
  })

  it('should work with `size` prop', async () => {
    ;(['small', 'medium', 'large'] as const).forEach((i) => {
      const wrapper = mount(NSelect, { props: { size: i } })
      expect(
        wrapper.find('.n-base-selection').attributes('style')
      ).toMatchSnapshot()
      wrapper.unmount()
    })
  })

  it('should work with `themeOverrides` prop', async () => {
    const selectThemeOverrides = {
      peers: {
        InternalSelectMenu: {
          peers: {
            Empty: {
              textColor: '#4fb233'
            }
          }
        }
      }
    }
    const wrapper = mount(NSelect, {
      props: {
        themeOverrides: selectThemeOverrides,
        show: true
      }
    })
    const menuWrapper = wrapper.findComponent(NInternalSelectMenu)
    expect(
      menuWrapper
        .find('.n-base-select-menu__empty .n-empty')
        .attributes('style')
    ).toContain('--n-text-color: #4fb233;')
    wrapper.unmount()
  })

  it('should work with `menuProps` prop', () => {
    const menuProps = {
      class: 'menu-test',
      style: {
        background: 'rgb(79, 178, 51)'
      }
    }
    const wrapper = mount(NSelect, {
      props: {
        menuProps,
        show: true
      }
    })
    const menuWrapper = wrapper.findComponent(NInternalSelectMenu)
    expect(menuWrapper.attributes('style')).toContain(
      'background: rgb(79, 178, 51);'
    )
    expect(menuWrapper.classes()).toContain('menu-test')
    wrapper.unmount()
  })

  it('should work with `action` slot', () => {
    const wrapper = mount(NSelect, {
      props: {
        show: true
      },
      slots: {
        action: () => 'test-action-slot'
      }
    })
    const menuWrapper = wrapper.findComponent(NInternalSelectMenu)
    expect(menuWrapper.find('.n-base-select-menu__action').exists()).toBe(true)
    expect(menuWrapper.find('.n-base-select-menu__action').text()).toContain(
      'test-action-slot'
    )
    wrapper.unmount()
  })

  it('should work with `empty` slot', () => {
    const wrapper = mount(NSelect, {
      props: {
        show: true
      },
      slots: {
        empty: () => 'test-empty-slot'
      }
    })
    const menuWrapper = wrapper.findComponent(NInternalSelectMenu)
    expect(menuWrapper.find('.n-base-select-menu__empty').exists()).toBe(true)
    expect(menuWrapper.find('.n-base-select-menu__empty').text()).toContain(
      'test-empty-slot'
    )
    wrapper.unmount()
  })
})
