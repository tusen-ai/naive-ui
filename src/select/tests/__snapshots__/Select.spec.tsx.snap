// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`n-select should work with \`size\` prop 1`] = `"--n-bezier: cubic-bezier(.4, 0, .2, 1); --n-border: 1px solid rgb(224, 224, 230); --n-border-active: 1px solid #18a058; --n-border-focus: 1px solid #36ad6a; --n-border-hover: 1px solid #36ad6a; --n-border-radius: 3px; --n-box-shadow-active: 0 0 0 2px rgba(24, 160, 88, 0.2); --n-box-shadow-focus: 0 0 0 2px rgba(24, 160, 88, 0.2); --n-box-shadow-hover: none; --n-caret-color: #18a058; --n-color: rgba(255, 255, 255, 1); --n-color-active: rgba(255, 255, 255, 1); --n-color-disabled: rgb(250, 250, 252); --n-font-size: 14px; --n-height: 28px; --n-padding-single-top: 0; --n-padding-multiple-top: 3px; --n-padding-single-right: 26px; --n-padding-multiple-right: 26px; --n-padding-single-left: 12px; --n-padding-multiple-left: 12px; --n-padding-single-bottom: 0; --n-padding-multiple-bottom: 0; --n-placeholder-color: rgba(194, 194, 194, 1); --n-placeholder-color-disabled: rgba(209, 209, 209, 1); --n-text-color: rgb(51, 54, 57); --n-text-color-disabled: rgba(194, 194, 194, 1); --n-arrow-color: rgba(194, 194, 194, 1); --n-arrow-color-disabled: rgba(209, 209, 209, 1); --n-loading-color: #18a058; --n-color-active-warning: rgba(255, 255, 255, 1); --n-box-shadow-focus-warning: 0 0 0 2px rgba(240, 160, 32, 0.2); --n-box-shadow-active-warning: 0 0 0 2px rgba(240, 160, 32, 0.2); --n-box-shadow-hover-warning: none; --n-border-warning: 1px solid #f0a020; --n-border-focus-warning: 1px solid #fcb040; --n-border-hover-warning: 1px solid #fcb040; --n-border-active-warning: 1px solid #f0a020; --n-color-active-error: rgba(255, 255, 255, 1); --n-box-shadow-focus-error: 0 0 0 2px rgba(208, 48, 80, 0.2); --n-box-shadow-active-error: 0 0 0 2px rgba(208, 48, 80, 0.2); --n-box-shadow-hover-error: none; --n-border-error: 1px solid #d03050; --n-border-focus-error: 1px solid #de576d; --n-border-hover-error: 1px solid #de576d; --n-border-active-error: 1px solid #d03050; --n-clear-size: 16px; --n-clear-color: rgba(194, 194, 194, 1); --n-clear-color-hover: rgba(146, 146, 146, 1); --n-clear-color-pressed: rgba(175, 175, 175, 1); --n-arrow-size: 16px;"`;

exports[`n-select should work with \`size\` prop 2`] = `"--n-bezier: cubic-bezier(.4, 0, .2, 1); --n-border: 1px solid rgb(224, 224, 230); --n-border-active: 1px solid #18a058; --n-border-focus: 1px solid #36ad6a; --n-border-hover: 1px solid #36ad6a; --n-border-radius: 3px; --n-box-shadow-active: 0 0 0 2px rgba(24, 160, 88, 0.2); --n-box-shadow-focus: 0 0 0 2px rgba(24, 160, 88, 0.2); --n-box-shadow-hover: none; --n-caret-color: #18a058; --n-color: rgba(255, 255, 255, 1); --n-color-active: rgba(255, 255, 255, 1); --n-color-disabled: rgb(250, 250, 252); --n-font-size: 14px; --n-height: 34px; --n-padding-single-top: 0; --n-padding-multiple-top: 3px; --n-padding-single-right: 26px; --n-padding-multiple-right: 26px; --n-padding-single-left: 12px; --n-padding-multiple-left: 12px; --n-padding-single-bottom: 0; --n-padding-multiple-bottom: 0; --n-placeholder-color: rgba(194, 194, 194, 1); --n-placeholder-color-disabled: rgba(209, 209, 209, 1); --n-text-color: rgb(51, 54, 57); --n-text-color-disabled: rgba(194, 194, 194, 1); --n-arrow-color: rgba(194, 194, 194, 1); --n-arrow-color-disabled: rgba(209, 209, 209, 1); --n-loading-color: #18a058; --n-color-active-warning: rgba(255, 255, 255, 1); --n-box-shadow-focus-warning: 0 0 0 2px rgba(240, 160, 32, 0.2); --n-box-shadow-active-warning: 0 0 0 2px rgba(240, 160, 32, 0.2); --n-box-shadow-hover-warning: none; --n-border-warning: 1px solid #f0a020; --n-border-focus-warning: 1px solid #fcb040; --n-border-hover-warning: 1px solid #fcb040; --n-border-active-warning: 1px solid #f0a020; --n-color-active-error: rgba(255, 255, 255, 1); --n-box-shadow-focus-error: 0 0 0 2px rgba(208, 48, 80, 0.2); --n-box-shadow-active-error: 0 0 0 2px rgba(208, 48, 80, 0.2); --n-box-shadow-hover-error: none; --n-border-error: 1px solid #d03050; --n-border-focus-error: 1px solid #de576d; --n-border-hover-error: 1px solid #de576d; --n-border-active-error: 1px solid #d03050; --n-clear-size: 16px; --n-clear-color: rgba(194, 194, 194, 1); --n-clear-color-hover: rgba(146, 146, 146, 1); --n-clear-color-pressed: rgba(175, 175, 175, 1); --n-arrow-size: 16px;"`;

exports[`n-select should work with \`size\` prop 3`] = `"--n-bezier: cubic-bezier(.4, 0, .2, 1); --n-border: 1px solid rgb(224, 224, 230); --n-border-active: 1px solid #18a058; --n-border-focus: 1px solid #36ad6a; --n-border-hover: 1px solid #36ad6a; --n-border-radius: 3px; --n-box-shadow-active: 0 0 0 2px rgba(24, 160, 88, 0.2); --n-box-shadow-focus: 0 0 0 2px rgba(24, 160, 88, 0.2); --n-box-shadow-hover: none; --n-caret-color: #18a058; --n-color: rgba(255, 255, 255, 1); --n-color-active: rgba(255, 255, 255, 1); --n-color-disabled: rgb(250, 250, 252); --n-font-size: 15px; --n-height: 40px; --n-padding-single-top: 0; --n-padding-multiple-top: 3px; --n-padding-single-right: 26px; --n-padding-multiple-right: 26px; --n-padding-single-left: 12px; --n-padding-multiple-left: 12px; --n-padding-single-bottom: 0; --n-padding-multiple-bottom: 0; --n-placeholder-color: rgba(194, 194, 194, 1); --n-placeholder-color-disabled: rgba(209, 209, 209, 1); --n-text-color: rgb(51, 54, 57); --n-text-color-disabled: rgba(194, 194, 194, 1); --n-arrow-color: rgba(194, 194, 194, 1); --n-arrow-color-disabled: rgba(209, 209, 209, 1); --n-loading-color: #18a058; --n-color-active-warning: rgba(255, 255, 255, 1); --n-box-shadow-focus-warning: 0 0 0 2px rgba(240, 160, 32, 0.2); --n-box-shadow-active-warning: 0 0 0 2px rgba(240, 160, 32, 0.2); --n-box-shadow-hover-warning: none; --n-border-warning: 1px solid #f0a020; --n-border-focus-warning: 1px solid #fcb040; --n-border-hover-warning: 1px solid #fcb040; --n-border-active-warning: 1px solid #f0a020; --n-color-active-error: rgba(255, 255, 255, 1); --n-box-shadow-focus-error: 0 0 0 2px rgba(208, 48, 80, 0.2); --n-box-shadow-active-error: 0 0 0 2px rgba(208, 48, 80, 0.2); --n-box-shadow-hover-error: none; --n-border-error: 1px solid #d03050; --n-border-focus-error: 1px solid #de576d; --n-border-hover-error: 1px solid #de576d; --n-border-active-error: 1px solid #d03050; --n-clear-size: 16px; --n-clear-color: rgba(194, 194, 194, 1); --n-clear-color-hover: rgba(146, 146, 146, 1); --n-clear-color-pressed: rgba(175, 175, 175, 1); --n-arrow-size: 16px;"`;
